src/tactic/polyrith.lean

```diff
@@ -0,0 +1,620 @@
+tactic succeeds. In other words, `linear_combination` is a certificate checker, and it is left
+to the user to find a collection of good coefficients. The `polyrith` tactic automates this
+process using the theory of Groebner bases.
+
+Polyrith does this by first parsing the relevant hypotheses into a form that Python can understand.
+It then calls a Python file that uses the SageMath API to compute the coefficients. These
+coefficients are then sent back to Lean, which parses them into pexprs. The information is then
+given to the `linear_combination` tactic, which completes the process by checking the certificate.
+
+## References
+
+* See the book `Ideals, Varieties, and Algorithms` by David Cox, John Little, and Donal O'Shea
+for the background theory on Groebner bases
+* This code was heavily inspired by the code for the tactic `linarith`, which was written by
+Robert Lewis, who advised me on this project as part of a Computer Science independant study
+at Brown University.
+
+-/
+
+open tactic native
+
+namespace polyrith
+
+/-! # Poly Datatype-/
+
+/--
+A datatype representing the semantics of multivariable polynomials.
+Each `poly` can be converted into a string.
+-/
+inductive poly
+| const: ℚ → poly
+| var: ℕ → poly
+| add: poly → poly → poly
+| sub: poly → poly → poly
+| mul: poly → poly → poly
+| pow : poly → ℕ → poly
+
+/--
+This converts a poly object into a string representing it. The string
+maintains the semantic structure of the poly object.
+-/
+
+meta def poly.mk_string : poly → string
+| (poly.const z) := to_string z
+| (poly.var n) := "var" ++ to_string n
+| (poly.add p q) := "(" ++ poly.mk_string p ++ " + " ++ poly.mk_string q ++ ")"
+| (poly.sub p q) := "(" ++ poly.mk_string p ++ " - " ++ poly.mk_string q ++ ")"
+| (poly.mul p q) := "(" ++ poly.mk_string p ++ " * " ++ poly.mk_string q ++ ")"
+| (poly.pow p n) := to_string $ format!"({poly.mk_string p} ^ {n})"
+
+meta instance : has_add poly := ⟨poly.add⟩
+meta instance : has_sub poly := ⟨poly.sub⟩
+meta instance : has_mul poly := ⟨poly.mul⟩
+meta instance : has_pow poly ℕ := ⟨poly.pow⟩
+meta instance : has_repr poly := ⟨poly.mk_string⟩
+meta instance : has_to_format poly := ⟨to_fmt ∘ poly.mk_string⟩
+meta instance : inhabited poly := ⟨poly.const 0⟩
+
+
+/-!
+# Parsing algorithms
+
+The following section contains code that can convert an `expr` of type `Prop` into a `poly` object
+(provided that it is an equality)
+-/
+
+local notation `exmap` := list (expr × ℕ)
+
+/--
+`poly_form_of_atom red map e` is the atomic case for `poly_form_of_expr`.
+If `e` appears with index `k` in `map`, it returns the singleton sum `poly.var k`.
+Otherwise it updates `map`, adding `e` with index `n`, and returns the singleton `poly.var n`.
+-/
+meta def poly_form_of_atom (red : transparency) (m : exmap) (e : expr) : tactic (exmap × poly) :=
+(do (_, k) ← m.find_defeq red e, return (m, poly.var k)) <|>
+(let n := m.length + 1 in return ((e, n)::m, poly.var n))
```

Comment:
It seems that this only ever stores list of the form `[(a, 0), (b,1), (c, 2), ...]`, which have a simpler representation as `[a, b, c, ...]`. You can achieve that with
```suggestion
/--
`poly_form_of_atom red vars e` is the atomic case for `poly_form_of_expr`.
If `e` appears with index `k` in `vars`, it returns the singleton sum `poly.var k`.
Otherwise it updates `vars`, adding `e` with index `n`, and returns the singleton `poly.var n`.
-/
meta def poly_form_of_atom (red : transparency) (vars : list expr) (e : expr) :
  tactic (list expr × poly) :=
do
  index_of_e ← vars.mfoldl_with_index
    (λ n last e', returnopt last <|> tactic.is_def_eq e e' red >> return (some n))
    none,
  match index_of_e with
  | some k := return (vars, poly.var k)
  | none   := return (e :: vars, poly.var vars.length)
  end
```

This is slightly longer, but it also removes the need for `list.find_defeq` and `exmap`.

Reply:
What's actually missing here is `list.mfind_index` to match `list.find_index`, if that looks too long to you

Reply:
Somehow I was sure that `find_defeq` was being added as part of this PR, but I guess it already exists in mathlib.